File: python/백준/그래프/BFS/11725 트리의 부모 찾기.py
import sys
from collections import deque

input = sys.stdin.readline
n = int(input())

graph = [[] for _ in range(n + 1)]
for _ in range(n - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
    
parents = [0] * (n +  1)

def bfs(start):
    queue = deque([start])

    while queue:
        node = queue.popleft()
    
        for neighbor in graph[node]:
            if parents[neighbor] == 0:
                parents[neighbor] = node
                queue.append(neighbor)
            
bfs(1)

for i in range(2, n + 1):
    print(parents[i])

File: cpp/백준/그래프/BFS/11725 트리의 부모 찾기.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<vector<int>> graph(n + 1);
    for (int i = 0; i < n - 1; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    vector<int> parents(n + 1, 0);

    queue<int> q;
    q.push(1);
    while (!q.empty()) {
        int node = q.front();
        q.pop();
        for (int neighbor : graph[node]) {
            if (parents[neighbor] == 0 && neighbor != 1) {
                parents[neighbor] = node;
                q.push(neighbor);
            }
        }
    }

    for (int i = 2; i <= n; ++i) cout << parents[i] << "\n";
    return 0;
}
